File: Data Structure/Union find/Medium/LET_no721_Accounts Merge.cpp
class Solution {
public:
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        vector<int> parent(accounts.size(), 0);
        map<string, int> email_id;
        for (int i = 0; i < accounts.size(); i++) {
            parent[i] = i;
            for (int j = 1; j < accounts[i].size(); j++) {
                auto it = email_id.find(accounts[i][j]);
                if (it != email_id.end()) {
                    int p1 = find(parent, i);
                    int p2 = find(parent, it->second);
                    parent[p1] = p2;
                }
                else
                    email_id[accounts[i][j]] = i;
            }
        }
        
        unordered_map<int, vector<string>> mres;
        for (auto& it : email_id) {
            int p = find(parent, it.second);
            mres[p].push_back(it.first);
        }
        
        vector<vector<string>> res;
        for (auto& it : mres) {
            vector<string> temp = {accounts[it.first][0]};
            for (auto& i : it.second)
                temp.push_back(i);
            res.push_back(temp);
        }
        return res;
    }
    
    int find(vector<int>& parent, int i) {
        if (parent[i] != i)
            parent[i] = find(parent, parent[i]);
        return parent[i];
    }
};

File: Data Structure/Union find/Medium/LET_no721_Accounts Merge.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.email_to_id = dict()
    
    def find(self, i):
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    
    def union(self, i, email):
        if email not in self.email_to_id:
            self.email_to_id[email] = i
            return
        
        j = self.email_to_id[email]
        pi, pj = self.find(i), self.find(j)
        if pi != pj:
            self.parent[pi] = pj

class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        n = len(accounts)
        uf = UnionFind(n)
        
        for i in range(n):
            for j in range(1, len(accounts[i])):
                uf.union(i, accounts[i][j])
        
        id_to_email = collections.defaultdict(list)
        for email, i in uf.email_to_id.items():
            p = uf.find(i)
            id_to_email[p].append(email)
            
        res = []
        for i, emails in id_to_email.items():
            temp = [accounts[i][0]]
            temp.extend(sorted(emails))
            res.append(temp)
        
        return res